edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java

```diff
@@ -143,20 +137,57 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
             return false;
         }
 
+        var assignedGroups = groups.getContent();
+
         //call evaluator function
         return OPERATOR_EVALUATOR_MAP.get(operator).apply(new BpnGroupHolder(assignedGroups, rightOperand));
     }
 
+    /**
+     * Verify if BPN was found in the database and if the BPN has any group assigned. If one of those validations fail,
+     * it will return an error message, otherwise it will return {@code null}.
+     * The evaluation is prematurely aborted (returns {@code null}) if:
+     * <ul>
+     *     <li>{@link Operator} is a negative operator. Check {@link BusinessPartnerGroupFunction#NEGATIVE_OPERATORS}</li>
+     * </ul>
+     *
+     * @param groups    result of the BPN search on the database
+     * @param operator  operator sent for evaluation
+     * @param bpn       participant agent identifier
+     * @return  an error message in case of empty groups or empty assigned groups, {@code null} otherwise.
+     */
+    private String evaluateEmptyGroups(StoreResult<List<String>> groups, Operator operator, String bpn) {
+        var isNegativeOperator = NEGATIVE_OPERATORS.contains(operator);
+
+        if (isNegativeOperator) {
+            return null;
+        }
+
+        // BPN not found in database
+        if (groups.failed()) {
```

Comment:
like I remarked above, `Result`s should be passed into and evaluate in methods. The code flow of this method and its caller can be simplified/unified and made more readable.